pkg/meta/distributed_lock.go

```diff
@@ -0,0 +1,364 @@
+	return json.Unmarshal(data, c)
+}
+
+// DistributedLock is a distributed lock implementation based on meta key transaction.
+type DistributedLock struct {
+	outerCtx   context.Context
+	store      DataStore
+	instanceID string
+
+	lockName []byte
+	backoff  time.Duration
+	lease    time.Duration
+
+	cancelRenewLoop context.CancelFunc
+}
+
+// DistributedLockBuilder is used to build a DistributedLock.
+type DistributedLockBuilder struct {
+	backoff time.Duration
+	lease   time.Duration
+}
+
+// NewDistributedLockBuilder creates a new DistributedLockBuilder.
+func NewDistributedLockBuilder() *DistributedLockBuilder {
+	return &DistributedLockBuilder{
+		backoff: 3 * time.Second,
+		lease:   30 * time.Second,
+	}
+}
+
+// SetBackoff sets the backoff time between two lock operations.
+// Default is 3 * time.Second.
+func (b *DistributedLockBuilder) SetBackoff(backoff time.Duration) *DistributedLockBuilder {
+	b.backoff = backoff
+	return b
+}
+
+// SetLease sets the lease time of the lock.
+// Default is 30 * time.Second. 0 means no lease.
+func (b *DistributedLockBuilder) SetLease(lease time.Duration) *DistributedLockBuilder {
+	b.lease = lease
+	return b
+}
+
+// Build builds a DistributedLock.
+func (b *DistributedLockBuilder) Build(
+	ctx context.Context, store DataStore, instanceID string, lockName string) *DistributedLock {
+	ctx = logutil.WithCategory(ctx, "distributed-lock")
+	return &DistributedLock{
+		outerCtx:   ctx,
+		store:      store,
+		instanceID: instanceID,
+
+		lockName: []byte(lockName),
+		backoff:  b.backoff,
+		lease:    b.lease,
+	}
+}
+
+func (d *DistributedLock) renewLockLoop(ctx context.Context) {
+	ticker := time.NewTicker(d.lease / 3)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := d.renew(ctx)
+			if err != nil {
+				logutil.Logger(d.outerCtx).Warn("renew lock failed", zap.Error(err))
+				return
+			}
+		}
+	}
+}
+
+func (d *DistributedLock) renew(ctx context.Context) error {
+	return d.getLockInTxnWithRetry(ctx,
+		func(lc *lockContent, txn DataTxn) (storeErr, otherErr error) {
+			if lc == nil || lc.InstanceID != d.instanceID {
+				logutil.Logger(d.outerCtx).Warn("lock is unexpectedly removed", zap.String("instanceID", d.instanceID))
+				return nil, nil
+			}
+
+			// Lock is exist, update it.
+			lc.StartTS = txn.StartTS()
+			data, err := lc.encode()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			err = txn.Set(d.lockName, data)
+			if err != nil {
+				return errors.Trace(err), nil
+			}
+			return nil, nil
+		})
+}
+
+var (
+	// lockHeldErr is returned when the lock is held by another routine.
+	lockHeldErr = errors.Errorf("lock is held")
+	// lockHeldNonBlockingErr is used to indicate that the lock is held and the caller should not retry.
+	lockHeldNonBlockingErr = errors.Errorf("lock is held, non-blocking")
+)
+
+// Lock locks the distributed lock.
+func (d *DistributedLock) Lock() error {
```

Comment:
For this implement, seems try lock and lock are almost the same. lock should stop and wait until unlock to release blocked applier.

Reply:
Yes. This is the difference between `Lock` and `TryLock`.